common/src/main/java/com/microsoft/identity/common/internal/controllers/NativeAuthMsalController.kt

```diff
@@ -508,6 +524,302 @@ class NativeAuthMsalController : BaseNativeAuthController() {
+            val currentTime = System.currentTimeMillis()
+            return currentTime - startTime > ResetPasswordSubmitNewPasswordCommand.POLL_COMPLETION_TIMEOUT_IN_MILISECONDS
+        }
+
+        fun pollIntervalIsAppropriate(pollIntervalInSeconds: Int?): Boolean {
+            return pollIntervalInSeconds != null && pollIntervalInSeconds <= 15 && pollIntervalInSeconds >= 1
+        }
+
+        val methodTag = "$TAG:resetPasswordPollCompletion"
+
+        LogSession.logMethodCall(TAG, "${TAG}.resetPasswordPollCompletion")
+
+        try {
+            val pollWaitInterval: Int = if (!pollIntervalIsAppropriate(pollIntervalInSeconds)) {
+                ResetPasswordSubmitNewPasswordCommand.DEFAULT_POLL_COMPLETION_INTERVAL_IN_MILISECONDS
+            } else {
+                pollIntervalInSeconds!! * 1000
+            }
+
+            var pollCompletionApiResult = performResetPasswordPollCompletionCall(
+                oAuth2Strategy = oAuth2Strategy,
+                passwordResetToken = passwordResetToken
+            )
+
+            val startTime = System.currentTimeMillis()
+
+            while (pollCompletionApiResult is ResetPasswordPollCompletionApiResult.InProgress) {
+                ThreadUtils.sleepSafely(
+                    pollWaitInterval,
+                    methodTag,
+                    "Waiting between reset password polls"
+                )
+
+                if (pollCompletionTimedOut(startTime)) {
+                    Logger.warn(
+                        TAG,
+                        "Reset password completion timed out."
+                    )
+                    return ResetPasswordCommandResult.PasswordResetFailed(
+                        error = ResetPasswordSubmitNewPasswordCommand.POLL_COMPLETION_TIMEOUT_ERROR_CODE,
+                        errorDescription = ResetPasswordSubmitNewPasswordCommand.POLL_COMPLETION_TIMEOUT_ERROR_DESCRIPTION
+                    )
+                }
+
+                pollCompletionApiResult = performResetPasswordPollCompletionCall(
+                    oAuth2Strategy = oAuth2Strategy,
+                    passwordResetToken = passwordResetToken
+                )
+            }
+
+            return when (pollCompletionApiResult) {
+                is ResetPasswordPollCompletionApiResult.PollingFailed -> {
+                    ResetPasswordCommandResult.PasswordResetFailed(
+                        error = pollCompletionApiResult.error,
+                        errorDescription = pollCompletionApiResult.errorDescription
+                    )
+                }
+
+                is ResetPasswordPollCompletionApiResult.PollingSucceeded -> {
+                    ResetPasswordCommandResult.Complete
+                }
+
+                is ResetPasswordPollCompletionApiResult.InProgress -> {
+                    Logger.warn(
+                        TAG,
+                        "in_progress received after polling, illegal state"
```

Comment:
I guess our own code threw this error? In that case can we handle the situation where it originated in our code?

Reply:
This `is` clause is present because `when` statement requires all clauses to be present. This log should not be printed.

Reply:
InProgress is a valid state because that's the state when we are sleeping and polling the server. We have to put the InProgress here because the compiler requires all possible states to be enumerated in the `when` statement. We can remove this and replace it with an `else` block (that will be acceptable to compiler). We have chosen to use the state name instead of `else` block because else should generally be used to represent multiple unhandled states.